Fobnail provisioning

arturkow2000 opened the pull request.

arturkow2000 committed:
pal_nrf: add module for detecting button presses

arturkow2000 committed:
Add basic module for Fobnail provisioning

Add skeleton for Fobnail Token provisioning module, configure Fobnail to
trigger re-provisioning after holding button for 10 secs.

arturkow2000 committed:
Detect whether Fobnail is provisioned, signal reprovisioning with leds

arturkow2000 committed:
pal_pc: add dummy button module to keep API in sync with pal_nrf

arturkow2000 committed:
Clear persistent storage when re-provisioning

When user triggers re-provisioning, Fobnail is reset into factory state
by erasing persistent storage region. After reboot Fobnail enters
provisioning mode.

arturkow2000 committed:
token provisioning: add state machine skeleton

arturkow2000 committed:
Improve state machine

Add missing calls: done() method, don't panic on error, instead re-try
provisioning.

krystian-hebel commented on src/main.rs:
```diff
@@ -62,10 +63,22 @@ where
 
+/// Checks whether we have any certificates installed in certstore. This is used
+/// to detect whether Fobnail has been provisioned.
+fn have_certchain<T>(trussed: &mut T) -> bool
```
According to today's discussion we probably won't keep whole chain saved, only certificates that will be created in 4.5. This function may work without changes to its code, but it should be renamed to something more appropriate.

krystian-hebel commented on src/main.rs:
```diff
@@ -151,15 +177,46 @@ fn main() -> ! {
     loop {
+        let now = pal::timer::get_time_ms() as u64;
+        // Pressing button for 10 s triggers Fobnail Token provisioning.
+        if now - button_press_time > 10000
+            && !matches!(operation_mode, OperationMode::TokenProvisioning(_))
```
This only checks if button state changed more than 10s ago, but doesn't check its current state. It probably will purge the device 10s after button is released, even when it was pressed for shorter period.

arturkow2000 replied:
Fixed

krystian-hebel commented on src/main.rs:
```diff
@@ -151,15 +177,46 @@ fn main() -> ! {
     loop {
+        let now = pal::timer::get_time_ms() as u64;
+        // Pressing button for 10 s triggers Fobnail Token provisioning.
+        if now - button_press_time > 10000
+            && !matches!(operation_mode, OperationMode::TokenProvisioning(_))
```
This only checks if button state changed more than 10s ago, but doesn't check its current state. It probably will purge the device 10s after button is released, even when it was pressed for shorter period.

krystian-hebel replied:
Duplicated comment, marking as resolved.

krystian-hebel commented on src/main.rs:
```diff
@@ -189,6 +246,12 @@ fn main() -> ! {
+            OperationMode::TokenProvisioning(ref mut p) => {
+                p.poll(coap_socket);
+                if p.done() {
+                    operation_mode = operation_mode.provisioning();
```
Local platform provisioning happens on a different device than Fobnail provisioning, so I think it should report success and enter indefinite idle state instead. This could change for remote platform provisioning but we're not implementing it yet.

arturkow2000 replied:
@krystian-hebel We can easily do that, however then we would have to get `rtthost` (or something else) working to re-connect RTT without reflashing firmware (`cargo-embed` doesn't allow that)

krystian-hebel replied:
Fair point.

krystian-hebel commented on src/main.rs:
```diff
@@ -189,6 +246,12 @@ fn main() -> ! {
+            OperationMode::TokenProvisioning(ref mut p) => {
+                p.poll(coap_socket);
+                if p.done() {
+                    operation_mode = operation_mode.provisioning();
```
Local platform provisioning happens on a different device than Fobnail provisioning, so I think it should report success and enter indefinite idle state instead. This could change for remote platform provisioning but we're not implementing it yet.

krystian-hebel replied:
Duplicated comment, marking as resolved.

krystian-hebel commented on src/main.rs:
```diff
@@ -80,11 +93,23 @@ impl<'a> OperationMode<'a> {
 
+    /// Enter Fobnail token provisioning mode.
+    pub fn token_provisioning(self) -> Self {
+        // FIXME: reinitializing CoapClient resets its message ID which is used
+        // for deduplication
```
We will have to do something about it in the long term, I think we are also missing some kind of acknowledge message to be sent from Fobnail to COAP server. Because of that server re-sends the same message (at least for fetch) because it thinks that the previous one didn't arrive to the client. It may pose an issue e.g. when multiple tokens are being provisioned one after another, without restarting the server in between.

krystian-hebel replied:
Duplicated comment, marking as resolved.

krystian-hebel commented on src/main.rs:
```diff
@@ -80,11 +93,23 @@ impl<'a> OperationMode<'a> {
 
+    /// Enter Fobnail token provisioning mode.
+    pub fn token_provisioning(self) -> Self {
+        // FIXME: reinitializing CoapClient resets its message ID which is used
+        // for deduplication
```
We will have to do something about it in the long term, I think we are also missing some kind of acknowledge message to be sent from Fobnail to COAP server. Because of that server re-sends the same message (at least for fetch) because it thinks that the previous one didn't arrive to the client. It may pose an issue e.g. when multiple tokens are being provisioned one after another, without restarting the server in between.

arturkow2000 replied:
@krystian-hebel missing ACK is highly probable as some ACK related code is missing. Last message ID can be transferred between clients when switching modes, when powering on Fobnail probably would be randomized.

krystian-hebel replied:
Please open an issue for this problem and we will deal with it later.

arturkow2000 replied:
https://github.com/fobnail/fobnail/issues/34

arturkow2000 committed:
Fix button press/release handling

Fixes problem triggering token provisioning mode after 10 sec even if
button was pressed for shorter period.

arturkow2000 committed:
Rename have_certchain to is_token_provisioned

We no longer save certificate chains in certstore. From now we will only
save a leaf certfificates. have_certchain name would be misleading.
